src/hotspot/share/nmt/regionsTree.hpp

```diff
@@ -0,0 +1,160 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#ifndef NMT_REGIONSTREE_HPP
+#define NMT_REGIONSTREE_HPP
+
+#include "logging/log.hpp"
+#include "nmt/nmtCommon.hpp"
+#include "nmt/vmatree.hpp"
+#include "nmt/virtualMemoryTracker.hpp"
+
+// RegionsTree extends VMATree to add some more specific API and also defines a helper
+// for processing the tree nodes in a shorter and more meaningful way.
+class RegionsTree : public VMATree {
+ private:
+  NativeCallStackStorage _ncs_storage;
+  bool _with_storage;
+
+ public:
+  RegionsTree(bool with_storage) : VMATree() , _ncs_storage(with_storage), _with_storage(with_storage) { }
+
+  ReservedMemoryRegion find_reserved_region(address addr);
+
+  SummaryDiff commit_region(address addr, size_t size, const NativeCallStack& stack);
+  SummaryDiff uncommit_region(address addr, size_t size);
+
+  using Node = VMATree::TreapNode;
+
+  class NodeHelper {
+      Node* _node;
+      public:
+      NodeHelper() : _node(nullptr) { }
+      NodeHelper(Node* node) : _node(node) { }
+      inline bool is_valid() { return _node != nullptr; }
+      inline void clear_node() { _node = nullptr; }
+      inline VMATree::position position() { return _node->key(); }
+      inline bool is_committed_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Committed) >= 2; }
+      inline bool is_released_begin() { return out_state() == VMATree::StateType::Released; }
+      inline bool is_reserved_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Reserved) == 1; }
+      inline VMATree::StateType in_state() { return _node->val().in.type(); }
+      inline VMATree::StateType out_state() { return _node->val().out.type(); }
+      inline size_t distance_from(NodeHelper& other) { return position() - other.position(); }
+      inline NativeCallStackStorage::StackIndex out_stack_index() { return _node->val().out.stack(); }
+      inline MemTag in_tag() { return _node->val().in.mem_tag(); }
+      inline MemTag out_tag() { return _node->val().out.mem_tag(); }
+      inline void set_in_tag(MemTag tag) { _node->val().in.set_tag(tag); }
+      inline void set_out_tag(MemTag tag) { _node->val().out.set_tag(tag); }
+      inline void print_on(outputStream* st) {
+        auto st_str = [&](int s){
+          return s == (int)VMATree::StateType::Released ? "Rl" :
+                 s ==  (int)VMATree::StateType::Reserved ? "Rv" : "Cm";
+        };
+        st->print_cr("pos: " INTPTR_FORMAT " "
+                     "%s, %s <|> %s, %s",
+                     p2i((address)position()),
+                     st_str((int)in_state()),
+                     NMTUtil::tag_to_name(in_tag()),
+                     st_str((int)out_state()),
+                     NMTUtil::tag_to_name(out_tag())
+                     );
+      }
+    };
```

Comment:
1. Doesn't need to be inline, move to `cpp` file.

2. No need to cast to `int`, just use the `VMATree::StateType` directly.

3. Should be wrapped in `#ifdef ASSERT` probably, I don't see us shipping this in product builds.

Reply:
Done.